docs/source/developers/entry_points_api/model_providers_group.md

```diff
@@ -0,0 +1,317 @@
+both in the chat UI and with magic commands. Otherwise, users can only use the new provider
+with magic commands.
+
+To make the new provider available, you need to declare it as an [entry point][entry_points]:
+
+```toml
+# my_package/pyproject.toml
+[project]
+name = "my_package"
+version = "0.0.1"
+
+[project.entry-points."jupyter_ai.model_providers"]
+my-provider = "my_provider:MyProvider"
+```
+
+To test that the above minimal provider package works, install it with:
+
+```sh
+# from `my_package` directory
+pip install -e .
+```
+
+Then, restart JupyterLab. You should now see an info message in the log that mentions
+your new provider's `id`:
+
+```
+[I 2023-10-29 13:56:16.915 AiExtension] Registered model provider `my_provider`.
+```
+
+### Adding custom API keys and fields
+
+You can add handle authentication via API keys, and configuration with
+custom parameters using an auth strategy and fields as shown in the example
+below.
+
+```python
+from typing import ClassVar, List
+from jupyter_ai_magics import BaseProvider
+from jupyter_ai_magics.providers import EnvAuthStrategy, Field, TextField, MultilineTextField
+from langchain_community.llms import FakeListLLM
+
+
+class MyProvider(BaseProvider, FakeListLLM):
+    id = "my_provider"
+    name = "My Provider"
+    model_id_key = "model"
+    models = [
+        "model_a",
+        "model_b"
+    ]
+
+    auth_strategy = EnvAuthStrategy(
+        name="MY_API_KEY", keyword_param="my_api_key_param"
+    )
+
+    fields: ClassVar[List[Field]] = [
+        TextField(key="my_llm_parameter", label="The name for my_llm_parameter to show in the UI"),
+        MultilineTextField(key="custom_config", label="Custom Json Config", format="json"),
+    ]
+
+    def __init__(self, **kwargs):
+        model = kwargs.get("model_id")
+        kwargs["responses"] = (
+            ["This is a response from model 'a'"]
+            if model == "model_a" else
+            ["This is a response from model 'b'"]
+        )
+        super().__init__(**kwargs)
+```
+
+The `auth_strategy` handles specifying API keys for providers and models.
+The example shows the `EnvAuthStrategy` which takes the API key from the
+environment variable with the name specified in `name` and be provided to the
+model's `__init__` as a kwarg with the name specified in `keyword_param`.
+This will also cause a field to be present in the configuration UI with the
+`name` of the environment variable as the label.
+
+Further configuration can be handled adding `fields` into the settings
+dialogue for your custom model by specifying a list of fields as shown in
+the example.  These will be passed into the `__init__` as kwargs, with the
+key specified by the key in the field object. The label specified in the field
+object determines the text shown in the configuration section of the user
+interface.
+
+## How-to: define custom code completions
+
+Any model provider derived from `BaseProvider` can be used as a completion provider.
```

Comment:
This and everything below is roughly duplicated in `inline_completions.md`. I don't know which copy you intended to keep but just highlighting this with assumption that only one was meant to be included.

Reply:
Great catch, thanks! I'll look into this when I'm able

Reply:
Removed in the latest commit. 👍